File: server/datasets/python_source-code/Multiple-C1/organic_9.py
"""
Graph BFS - Organic Submission #9
Functional-style implementation with type hints and docstrings.
"""

from collections import deque
from typing import Dict, List, Hashable


def bfs_traversal(graph: Dict[Hashable, List[Hashable]], source: Hashable) -> List[Hashable]:
    """Return the breadth-first traversal order starting at `source`."""
    frontier = deque([source])
    explored = {source}
    sequence: List[Hashable] = []

    while frontier:
        vertex = frontier.popleft()
        sequence.append(vertex)

        for neighbour in graph.get(vertex, []):
            if neighbour not in explored:
                explored.add(neighbour)
                frontier.append(neighbour)

    return sequence


def main() -> None:
    sample_graph = {
        "S": ["A", "B"],
        "A": ["S", "C"],
        "B": ["S", "C", "D"],
        "C": ["A", "B", "D"],
        "D": ["B", "C"],
    }
    print(bfs_traversal(sample_graph, "S"))


if __name__ == "__main__":
    main()

File: server/datasets/controlled_benchmarks/java/multiple_files/multiple_c1/organic_9.java
/*
 * Graph BFS - Organic Submission #9
 * Functional-style implementation with explicit typed collections.
 */

import java.util.*;

class Solution {
    public static List<String> bfsTraversal(Map<String, List<String>> graph, String source) {
        Queue<String> frontier = new ArrayDeque<>();
        Set<String> explored = new HashSet<>();
        List<String> sequence = new ArrayList<>();

        frontier.add(source);
        explored.add(source);

        while (!frontier.isEmpty()) {
            String vertex = frontier.poll();
            sequence.add(vertex);

            List<String> neighbours = graph.getOrDefault(vertex, Collections.emptyList());
            for (String neighbour : neighbours) {
                if (!explored.contains(neighbour)) {
                    explored.add(neighbour);
                    frontier.add(neighbour);
                }
            }
        }
        return sequence;
    }

    public static void main(String[] args) {
        Map<String, List<String>> sampleGraph = new HashMap<>();
        sampleGraph.put("S", Arrays.asList("A", "B"));
        sampleGraph.put("A", Arrays.asList("S", "C"));
        sampleGraph.put("B", Arrays.asList("S", "C", "D"));
        sampleGraph.put("C", Arrays.asList("A", "B", "D"));
        sampleGraph.put("D", Arrays.asList("B", "C"));

        System.out.println(bfsTraversal(sampleGraph, "S"));
    }
}
